src/librustdoc/visit_ast.rs

```diff
@@ -213,8 +215,36 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
 
     }
 
-    fn resolve_id(&mut self, id: ast::NodeId, renamed: Option<ast::Name>,
+    /// Tries to resolve the target of a `pub use` statement and inlines the
+    /// target if it is defined locally and would not be documented otherwise,
+    /// (due to the `strip-private` or `strip-hidden` passes)
+    /// or when it is specifically requested with `please_inline`.
+    /// (the latter is the case when the import is marked `doc(inline)`)
+    ///
+    /// Cross-crate inlining occurs later on during crate cleaning
+    /// and follows different rules.
+    ///
+    /// Returns true if the target has been inlined.
+    fn maybe_inline_local(&mut self, id: ast::NodeId, renamed: Option<ast::Name>,
                   glob: bool, om: &mut Module, please_inline: bool) -> bool {
+
+        fn is_doc_hidden(cx: &core::DocContext, mut node: ast::NodeId) -> bool {
+            let mut seen = HashSet::new();
+            seen.insert(node);
+
+            while let Some(id) = cx.map.get_enclosing_scope(node) {
+                node = id;
+                if !seen.insert(id) {
+                    break;
```

Comment:
I'm not quite sure why this is necessary. (`rustdoc/viewpath-self.rs` is caught in an infinite loop otherwise)